renderdoc/driver/gl/wrappers/gl_texture_funcs.cpp

```diff
@@ -1388,55 +1393,121 @@ void WrappedOpenGL::glCopyImageSubData(GLuint srcName, GLenum srcTarget, GLint s
+
+          uint32_t size =
+              (uint32_t)GetByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth, srcFmt, srcType);
+          srcCd.resize(size);
 
-          size_t srcSliceSize = GetCompressedByteSize(srcLevelWidth, srcLevelHeight,
-                                                      srcBlockSize[2], srcData.internalFormat);
-          size_t dstSliceSize = GetCompressedByteSize(dstLevelWidth, dstLevelHeight,
-                                                      dstBlockSize[2], dstData.internalFormat);
+          GLuint prevTex = 0;
+          GL.glGetIntegerv(srcData.curType == eGL_TEXTURE_2D ? eGL_TEXTURE_BINDING_2D
+                                                             : eGL_TEXTURE_BINDING_2D_ARRAY,
+                           (GLint *)&prevTex);
+
+          GLenum oldActive = eGL_TEXTURE0;
+          GL.glGetIntegerv(eGL_ACTIVE_TEXTURE, (GLint *)&oldActive);
+          GL.glActiveTexture(eGL_TEXTURE0);
+
+          GL.glBindTexture(srcData.curType, srcName);
+          GL.glGetTexImage(srcData.curType, srcLevel, srcFmt, srcType, srcCd.data());
+
+          GL.glBindTexture(srcData.curType, prevTex);
+          GL.glActiveTexture(oldActive);
+        }
+        if(!srcCd.isEmpty())
+        {
+          rdcfixedarray<uint32_t, 3> srcBlockSize =
+              srcIsCompressed ? GetCompressedBlockSize(srcData.internalFormat)
+                              : rdcfixedarray<uint32_t, 3>{1u, 1u, 1u};
+          rdcfixedarray<uint32_t, 3> dstBlockSize = GetCompressedBlockSize(dstData.internalFormat);
 
           RDCASSERT(srcWidth % srcBlockSize[0] == 0);
-          RDCASSERT(srcWidth % dstBlockSize[0] == 0);
           RDCASSERT(srcHeight % srcBlockSize[1] == 0);
-          RDCASSERT(srcHeight % dstBlockSize[1] == 0);
           RDCASSERT(srcDepth % srcBlockSize[2] == 0);
-          RDCASSERT(srcDepth % dstBlockSize[2] == 0);
-          for(size_t z = 0; z < (size_t)srcDepth; z += srcBlockSize[2])
+          RDCASSERT(srcX % srcBlockSize[0] == 0);
+          RDCASSERT(srcY % srcBlockSize[1] == 0);
+          RDCASSERT(srcZ % srcBlockSize[2] == 0);
+          RDCASSERT(dstX % dstBlockSize[0] == 0);
+          RDCASSERT(dstY % dstBlockSize[1] == 0);
+          RDCASSERT(dstZ % dstBlockSize[2] == 0);
+
+          size_t srcSize =
+              srcIsCompressed
+                  ? GetCompressedByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth,
+                                          srcData.internalFormat)
+                  : GetByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth, srcFmt, srcType);
+          size_t dstSize = GetCompressedByteSize(dstLevelWidth, dstLevelHeight, dstLevelDepth,
+                                                 dstData.internalFormat);
+
+          if(srcX == 0 && srcY == 0 && srcZ == 0 && dstX == 0 && dstY == 0 && dstZ == 0 &&
+             srcLevelWidth == srcWidth && srcLevelHeight == srcHeight &&
+             srcLevelDepth == srcDepth && srcSize == dstSize)
```

Comment:
I don't think this check is accurate anymore, since unless I'm mistaken the dimensions of the uncompressed texture will be off by a factor of the block size from the compressed texture - each uncompressed 1x1 texel maps to a compressed 4x4 block.

Reply:
It could be like 1x1 RGBA32UI -> 4x4 ETC2, which is in fact the case we use (virtual texture like and runtime gpu compression). So I just leave this fast code path when perform full texture copy of the same size.

Reply:
Yeh that's what I mean. In fact I didn't notice that the check had been changed, and it's now comparing `srcLevelWidth` to `srcWidth` which doesn't seem correct. I think the check for a full copy should be something like comparing src dimensions divided by src block size to dst dimensions divided by dst block size. Then it does the same kind of check as before (where it just compared src to dst) but with handling for compressed textures only being on one side.

Reply:
This fast path is ensured under these situations: 
- copy full srcTex (`srcX == 0 && srcY == 0 && srcZ == 0` and `srcLevelWidth == srcWidth && srcLevelHeight == srcHeight && srcLevelDepth == srcDepth`
- copy full dstTex (`dstX == 0 && dstY == 0 && dstZ == 0` and `srcSize == dstSize`)


>  src dimensions divided by src block size to dst dimensions divided by dst block size

I didn't check this way because it is implied by `srcSize == dstSize`